### PR TITLE
Add experimental extension for mutable command lists

### LINKED ISSUE #260: Experimental extension to mutate commands in a command list
# Summary
Introduce an experimental extension to mutate commands in a command list.

# Details

## Motivation
In order to efficiently support certain use-cases, such as compute graphs with dynamic parameters, it is desirable to have the ability to mutate commands within a recorded command list.
 
## Interoperability with Other Extensions
- This extension is complimentary to and may be used in conjunction with **ZE\_experimental\_command\_list\_clone**

### New Enumerations

**ze\_mutable\_command\_id\_exp\_flags\_t**
- mutable command identifier flags

**ze\_mutable\_command\_list\_exp\_flags\_t**
- mutable command list flags

### New Structures

**ze\_mutable\_command\_id\_exp\_desc\_t**
- Mutable command identifier descriptor

**ze\_mutable\_command\_list\_exp\_desc\_t**
- Mutable command list descriptor

**ze\_mutable\_commands\_exp\_desc\_t**
- Mutable commands descriptor

**ze\_mutable\_kernel\_argument\_exp\_desc\_t**
- Mutable kernel argument descriptor

**ze\_mutable\_group\_count\_exp\_desc\_t**
- Mutable kernel group count descriptor

**ze\_mutable\_group\_size\_exp\_desc\_t**
- Mutable kernel group size descriptor

**ze\_mutable\_global\_offset\_exp\_desc\_t**
- Mutable kernel global offset descriptor

#### zeCommandListGetNextCommandIdExp
This function returns a unique command identifier for the next command to be appended to a command list.

```cpp
ze_result_t zeCommandListGetNextCommandIdExp(
    ze_command_list_handle_t hCommandList,
    const ze_mutable_command_id_exp_desc_t* desc,
    uint64_t* pCommandId
);
```

| Parameter  | Description |
| :---------- | :---------------------------------------- |
| hCommandList | \[ ${\textsf{\color{orange}in}}$ \] handle of the command list |
| desc | \[ ${\textsf{\color{orange}in}}$ \] pointer to mutable command identifier descriptor |
| pCommandId | \[ ${\textsf{\color{yellow}out}}$ \] pointer to mutable command identifier to be written |

##### Notes
- this function may only be called for a mutable command list
- this function may not be called on a closed command list

---

#### zeCommandListUpdateMutableCommandsExp
This function updates mutable commands in a mutable command list.

```cpp
ze_result_t zeCommandListUpdateMutableCommandsExp(
    ze_command_list_handle_t hCommandList,
    const ze_mutable_commands_exp_desc_t* desc
);
```

| Parameter  | Description |
| :---------- | :---------------------------------------- |
| hCommandList | \[ ${\textsf{\color{orange}in}}$ \] handle of the command list |
| desc | \[ ${\textsf{\color{orange}in}}$ \] pointer to mutable commands descriptor; multiple descriptors may be chained via `pNext` member |

##### Notes
- this function may only be called for a mutable command list
- the application must synchronize a mutable command list before calling this function
- the application must close a mutable command list after completing all updates

---

#### zeCommandListUpdateMutableCommandSignalEventExp
This function updates the signal event for a mutable command in a mutable command list.

```cpp
ze_result_t zeCommandListUpdateMutableCommandSignalEventExp(
    ze_command_list_handle_t hCommandList,
    uint64_t commandId,
    ze_event_handle_t hSignalEvent
);
```

| Parameter  | Description |
| :---------- | :---------------------------------------- |
| hCommandList | \[ ${\textsf{\color{orange}in}}$ \] handle of the command list |
| commandId | \[ ${\textsf{\color{orange}in}}$ \] command identifier |
| hSignalEvent | \[ ${\textsf{\color{orange}in}}$ \]\[ ${\textsf{\color{lightblue}optional}}$ ] handle of the event to signal on completion |

##### Notes
- this function may only be called for a mutable command list
- the type, scope and flags of the signal event must match those of the source command
- passing a null pointer as the signal event will update the command to not issue a signal
- the application must synchronize a mutable command list before calling this function
- the application must close a mutable command list after completing all updates

---

#### zeCommandListUpdateMutableCommandWaitEventsExp
This function updates the wait events for a mutable command in a mutable command list.

```cpp
ze_result_t zeCommandListUpdateMutableCommandWaitEventsExp(
    ze_command_list_handle_t hCommandList,
    uint64_t commandId,
    ze_event_handle_t* phWaitEvents
);
```

| Parameter  | Description |
| :---------- | :---------------------------------------- |
|hCommandList | \[ ${\textsf{\color{orange}in}}$ \] handle of the command list |
|commandId | \[ ${\textsf{\color{orange}in}}$ \] command identifier |
|phWaitEvents | \[ ${\textsf{\color{orange}in}}$ \]\[ ${\textsf{\color{lightblue}optional}}$ \]\[ ${\textsf{\color{lightgreen}range\(0, numWaitEvents\)}}$ \] handle of the events to wait on before launching |

##### Notes
- this function may only be called for a mutable command list
- the number of wait events must match that of the source command
- the type, scope and flags of the wait events must match those of the source command
- passing `nullptr` as the wait events will update the command to not wait on any events prior to dispatch
- passing `nullptr` as an event on event wait list will remove event dependency from this wait list slot
- the application must synchronize a mutable command list before calling this function
- the application must close a mutable command list after completing all updates

---

#### Creating a Mutable Command List

```cpp
// Create a mutable command list
ze_mutable_command_list_exp_desc_t mutCmdListDesc = {
    ZE_STRUCTURE_TYPE_MUTABLE_COMMAND_LIST_EXP_DESC,
    nullptr,
    0 // flags
};

ze_command_list_desc_t commandListDesc = {
    ZE_STRUCTURE_TYPE_COMMAND_LIST_DESC,
    &mutCmdListDesc,
    0,
    0 // flags
};

ze_command_list_handle_t hCommandList = nullptr;
zeCommandListCreate(hContext, hDevice, &commandListDesc, &hCommandList);

// [ ...create fence, signal event and wait event objects... ]
// This example assumes hFence, hSignalEvent and hWaitEvent have been created.

// Create kernel from module
ze_kernel_desc_t kernelDesc = {
    ZE_STRUCTURE_TYPE_KERNEL_DESC,
    0,
    "example"
};
ze_kernel_handle_t hKernel = nullptr;
ze_result_t result = zeKernelCreate(hModule, &kernelDesc, &hKernel);

// Set the kernel arguments
ze_group_count_t groupSize = {};
zeKernelSuggestGroupSize(hKernel, 1024, 1024, 1, &groupSize.groupCountX, &groupSize.groupCountY, &groupSize.groupCountZ);

int defaultValue = 0;
zeKernelSetArgumentValue(hKernel, 0, sizeof(int), &defaultValue);

// Get next command identifier
ze_mutable_command_id_exp_desc_t cmdIdDesc = {
    ZE_STRUCTURE_TYPE_MUTABLE_COMMAND_ID_EXP_DESC       // stype
    nullptr,                                            // pNext
    0                                                   // flags
};

uint64_t commandId = 0;
zeCommandListGetNextCommandIdExp(hCommandList, &cmdIdDesc, &commandId);

// Encode command into command list
zeCommandListAppendLaunchKernel(hCommandList, hKernel, &groupSize, hSignalEvent, 1, &hWaitEvent);

// Close the command list
zeCommandListClose(hCommandList);

// ...

// Execute the command list
zeCommandQueueExecuteCommandLists(hCommandQueue, 1, &hCommandList, hFence);

// ...
```

#### Mutating Commands and Events

```cpp
// Modify group count
ze_group_count_t groupCount = {
    256,                                                    // groupCountX
    256,                                                    // groupCountY
    1                                                       // groupCountZ
};

ze_mutable_group_count_exp_desc_t groupCountDesc = {
    ZE_STRUCTURE_TYPE_MUTABLE_GROUP_COUNT_EXP_DESC,         // stype
    nullptr,                                                // pNext
    commandId,                                              // commandId
    &groupCount                                             // pGroupCount
};

// Modify Kernel Argument
int argValue = 1;

ze_mutable_kernel_argument_exp_desc_t krnlArgDesc = {
    ZE_STRUCTURE_TYPE_MUTABLE_KERNEL_ARGUMENT_EXP_DESC,     // stype
    &groupCountDesc,                                        // pNext
    commandId,                                              // commandId
    0,                                                      // argIndex
    sizeof(int),                                            // argSize
    &argValue                                               // pArgValue
};

// Prepare to update all mutable commands
ze_mutable_commands_exp_desc_t desc = {
    ZE_STRUCTURE_TYPE_MUTABLE_COMMANDS_EXP_DESC,            // stype
    &krnlArgDesc,                                           // pNext
    0                                                       // flags
};

// Synchronize the mutable command list
zeFenceHostSynchronize(hFence, UINT64_MAX);

// Update all mutable commands
zeCommandListUpdateMutableCommandsExp(hCommandList, &desc);

// Update signal event for the launch kernel command
zeCommandListUpdateMutableCommandSignalEventExp(hCommandList, commandId, hNewLaunchKernelSignalEvent);

// Update the wait event list for the launch kernel command
// Note, this example uses only a single wait event for the kernel
zeCommandListUpdateMutableCommandWaitEventsExp(hCommandList, commandId, &hNewLaunchKernelWaitEvent);

// Close the command list
zeCommandListClose(hCommandList);

// ...
```

**NOTE**: the command list must be explicitly closed after updating mutable commands and events

### PR DESCRIPTION
**Pull request by wdamon-intel**
Resolves #260

### CODE REVIEW

**Review by EwanC — APPROVED** [review]

**Review by MichalMrozek — APPROVED** [review]

File: scripts/core/mutableCommandList.yml [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+version: "1.9"
+name: $x_mutable_command_list_exp_version_t
+etors:
+    - name: "1_0"
+      value: "$X_MAKE_VERSION( 1, 0 )"
+      desc: "version 1.0"
+--- #--------------------------------------------------------------------------
+type: enum
+desc: "Supported mutable command identifier flags"
+class: $xCommandList
```
Comment (aravindksg) [inline scripts/core/mutableCommandList.yml:30]: version for this enum?

File: scripts/core/mutableCommandList.yml [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    - name: GLOBAL_OFFSET
+      desc: "kernel global offset"
+    - name: SIGNAL_EVENT
+      desc: "command signal event"
+    - name: WAIT_EVENTS
+      desc: "command wait events"
+--- #--------------------------------------------------------------------------
+type: struct
+desc: "Mutable command identifier descriptor"
+class: $xCommandList
```
Comment (aravindksg) [inline scripts/core/mutableCommandList.yml:48]: version

File: scripts/core/mutableCommandList.yml [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+members:
+    - type: $x_mutable_command_id_exp_flags_t
+      name: flags
+      desc: |
+            [in] mutable command identifier flags.
+              - must be 0 (default, equivalent to setting all flags), or a valid combination of $x_mutable_command_id_exp_flag_t
+--- #--------------------------------------------------------------------------
+type: enum
+desc: "Mutable command list flags"
+class: $xCommandList
```
Comment (aravindksg) [inline scripts/core/mutableCommandList.yml:60]: version

File: scripts/core/mutableCommandList.yml [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+members:
+    - type: $x_mutable_command_list_exp_flags_t
+      name: flags
+      desc: |
+            [in] mutable command list flags.
+              - must be 0 (default) or a valid combination of $x_mutable_command_list_exp_flag_t
+--- #--------------------------------------------------------------------------
+type: struct
+desc: "Mutable commands descriptor"
+class: $xCommandList
```
Comment (aravindksg) [inline scripts/core/mutableCommandList.yml:80]: version for this as well and for all the interfaces below

**Review by aravindksg — APPROVED** [review]

**Comment by wdamon-intel** [comment]
@zzdanowicz for review

File: scripts/core/EXT_Exp_MutableCommandList.rst [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    // Discover mutable command list properties
+    ${x}_mutable_command_list_exp_properties_t mutCmdListProps = {
+        ${X}_STRUCTURE_TYPE_MUTABLE_COMMAND_LIST_EXP_PROPERTIES,      // stype
+        nullptr,                                                    // pNext
+        0,                                                          // mutableCommandListFlags
+        0                                                           // mutableCommandFlags
+    };
+
+    ${x}_driver_properties_t driverProps = {
+        ${X}_STRUCTURE_TYPE_DEVICE_PROPERTIES
```
Comment (zzdanowicz) [inline scripts/core/EXT_Exp_MutableCommandList.rst:63]: DRIVER for `ZE_STRUCTURE_TYPE_DRIVER_PROPERTIES`

Comment (wdamon-intel) [inline scripts/core/EXT_Exp_MutableCommandList.rst:63]:
```suggestion
        ${X}_STRUCTURE_TYPE_DRIVER_PROPERTIES
```

**Review by MichalMrozek — APPROVED** [review]

File: scripts/core/mutableCommandList.yml [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+details:
+    - "This function may only be called for a mutable command list."
+    - "This function may not be called on a closed command list."
+    - "This function may be called from simultaneous threads with the same command list handle."
+    - "The implementation of this function should be lock-free."
+params:
+    - type: $x_command_list_handle_t
+      name: hCommandList
+      desc: "[in] handle of the command list"
+    - type: "const $x_mutable_command_id_exp_desc_t*"
```
Comment (zzdanowicz) [inline scripts/core/mutableCommandList.yml:195]: part of `x_mutable_command_id_exp_desc_t` is `x_mutable_command_list_exp_flags_t` - does it mean we can specify in advance when getting CommandID which part of next append are going to mutate?

Comment (zzdanowicz) [reply scripts/core/mutableCommandList.yml:195]: ok, thanks

Comment (wdamon-intel) [reply scripts/core/mutableCommandList.yml:195]: @zzdanowicz The mutable command id descriptor does not currently contain the command list flags, only the command flags. The command list _properties_ reports out both command list and command flags.

Is it a requirement that the command _list_ flags be built into the command id descriptor? The command flags should be sufficient.

Comment (zzdanowicz) [reply scripts/core/mutableCommandList.yml:195]: Yes, *command* flags. I just added (embargo) definitions, for the development purposes before official header arrive.
Treat them as temporary and only for sake of development - did I got it right?
```
typedef uint32_t ze_mutable_command_exp_flags_t;
typedef enum _ze_mutable_command_exp_flag_t {
    ZE_MUTABLE_COMMAND_EXP_FLAG_KERNEL_ARGUMENTS = ZE_BIT(0),
    ZE_MUTABLE_COMMAND_EXP_FLAG_GROUP_COUNT = ZE_BIT(1),
    ZE_MUTABLE_COMMAND_EXP_FLAG_GROUP_SIZE = ZE_BIT(2),
    ZE_MUTABLE_COMMAND_EXP_FLAG_GLOBAL_OFFSET = ZE_BIT(3),
    ZE_MUTABLE_COMMAND_EXP_FLAG_SIGNAL_EVENT = ZE_BIT(4),
    ZE_MUTABLE_COMMAND_EXP_FLAG_WAIT_EVENTS = ZE_BIT(5),
    ZE_MUTABLE_COMMAND_EXP_FLAG_FORCE_UINT32 = 0x7fffffff
} ze_mutable_command_exp_flag_t;

#define ZE_STRUCTURE_TYPE_MUTABLE_COMMAND_ID_EXP_DESC (ze_structure_type_t)0x00040001

typedef struct _ze_mutable_command_id_exp_desc_t {
    ze_structure_type_t stype;
    const void *pNext;
    ze_mutable_command_exp_flags_t flags;
} ze_mutable_command_id_exp_desc_t;
```

**Pull request merged**